### DISCUSSION TITLE
Smart quotes in Text that came from a string

### DISCUSSION THREAD

**Discussion by ghost** [typst/typst#1652] [Q&A] [answered]
Hi, I would like to use smart quotes also in  the text that came from a text string.
The reason is that the text is user data that is not necessarily in Typst syntax, so I can't just put it in `[...]`.
Is there a way to achieve this?

Here is the example code:
```typst
#set text(lang: "en")
#set par(justify: true)
- "Hello" :)
- #{["Hello"]} :)
- #{`"Hello"`.text} :(
- #{"\"Hello\""} :(
- #text(["Hello"]) :)
- #text("\"Hello\"") :(
```
![hello](https://github.com/typst/typst/assets/138529922/e55414b6-92bf-477d-b710-a509fdc5e846)

**Comment by tretre91** [comment, marked as answer]
You can use a show rule and the `smartquote()` function
```
#set text(lang: "en")
#set par(justify: true)
#show "\"": smartquote()
- "Hello" :)
- #{["Hello"]} :)
- #{`"Hello"`.text} :(
- #{"\"Hello\""} :(
- #text(["Hello"]) :)
- #text("\"Hello\"") :(
```
gives
![image](https://github.com/typst/typst/assets/63788850/1216906b-2332-4dee-93ce-dc1ce51be6a9)

**Reply by ghost** [reply]
Thanks, that's already quite good. Is it also possible to achieve this for `'`?
The following gives different results for string and content when `'` is used instead of `"`:

```typst
#set text(lang: "fr")
#set par(justify: true)
#show "\"": smartquote()
#show "'": smartquote()
- 'Hello' :)
- #{["Hello"]} :)
- #{`"Hello"`.text} :(
- #{"\"Hello\""} :(
- #text(["Hello"]) :)
- #text("'Hello'") :(
```

**Reply by ghost** [reply]
Ok, a workaround is to use for `'`: `#show "'": smartquote(double: false)`. That won't solve nested quotes, but it's good enough now, thanks!

**Reply by tretre91** [reply]
You're welcome !

> That won't solve nested quotes

It seems to work fine with nested quotes (assuming this is what you meant by nested quotes)

```
#set text(lang: "fr")
#set par(justify: true)
#show "\"": smartquote()
#show "'": smartquote(double: false)
#text("\"This 'should also \"work\"' for \"nested\" quotes\"")
```
![image](https://github.com/typst/typst/assets/63788850/8c7b6b44-1b4f-4c5e-95b4-b93a02ce5551)

**Reply by ghost** [reply]
I thought that smart quotes are smart enough to automatically detect when outter and inner quotes make sense.
Explicitly setting `double: true/false` requires to trust that the input author has already used the correct order.
So my hope was that the following was automatically turned into:

1. double
2. single
3. single
4. single
5. single
6. double

```typst
#set text(lang: "fr")
#set par(justify: true)
#show "\"": smartquote()
#show "'": smartquote()
#text("'Are \"smart\" 'quotes' smart?'")
```

But it turns out as
![smart](https://github.com/typst/typst/assets/138529922/0cca2dca-26bc-4abe-a217-078811bf6f72)

**Reply by tretre91** [reply]
Oh ok, I see what you mean. Yeah, smartquotes() cannot handle this afaik
